FW/Presentations/Lab_Guides/_L_C_04_Writing_UEFI_App_Simics_Linux_Lab.md

```diff
@@ -0,0 +1,2024 @@
+
+```bash
+$> cd ~/fw/edk2-ws/edk2-platforms/Platform/Intel/
+$> python build_bios.py -p BoardX58Ich10 GCC5
+```
+2. Copy SampleApp.efi from the build directory to the VHD Disk
+
+```bash
+$> cp ../Build/SimicsOpenBoardPkg/BoardX58Ich10/DEBUG_GCC5/SampleApp.efi ~/VHD
+```
+
+3. Run the qsp-modern-core script from Terminal Command Prompt
+
+```bash
+$>  ./simics targets/qsp-x86/qsp-modern-core.simics
+simics> run
+```
+
+4. At the UEFI Shell prompt
+
+```
+Shell> Fs1:
+FS1:/> SampleApp.efi
+System Table: 0x0DEFED018
+Press any key to continue:
+```
+
+5. Exit Simics
+
+```
+simics> stop
+simics> quit
+```
+
+---
+## Slide 55 Lab 5: Creating a Simple Typewriter Function
+
+In this lab, you'll learn how to create a simple typewriter function that retrieves the keys you type and subsequently prints each one back to the console
+
+---
+## Slide 56 Lab 5 : Typewriter Function
+
+Create a simple Typewriter Function using the SampleApp from Lab 4
+
+**Requirements:**
+
+- If the Typewriter Function is enabled
+- Retrieve keys entered from keyboard (like Lab 4)
+- Print back each key entered to the console
+- To exit, press "." (DOT) and then \<Enter\>
+
+```
+Shell> fs1:
+FS1:\> SampleApp
+System Table: 0xDEFED018
+
+Press any Key to continue :
+Enter text. Include a dot ('.') in a sentence then <Enter> to exit:
+this is the first line of the typewriter feature function.
+FS1:\>
+```
+
+---
+## Slide 57 Lab 5 : Typerwriter Function
+
+Create a simple Typewriter Function using the SampleApp from Lab 4
+
+**How:**
+
+1. Create a Feature Flag PCD to Enable
+2. Add a Loop using `WaitForEvent` with `WaitForKey`
+3. Use the `ReadKeyStroke` function from ConIn
+4. Print back each key to console
+5. Exit when DOT "." character is followed by an \<Enter\> key
+
+---
+## Slide 58 Lab 5: How Process (Hints)
+
+- Refer to Lab 2 for how PCDs work and create a PCD in the Platform DEC file.
+- Use the same procedure as with Lab 4 to find `ReadKeyStroke` in the workspace: [MdePkg/Library/UefiLib/Console.c](https://github.com/tianocore/edk2/blob/master/MdePkg/Library/UefiLib/Console.c) ~ In 552
+
+```c
+Status = gST->ConIn->ReadKeyStroke (gST->ConIn, Key);
+```
+
+- `ReadKeyStroke` uses a buffer called `EFI_INPUT_KEY` ~ In 393
+
+```
+OUT EFI_INPUT_KEY *Key,
+```
+
+- TIP: Good idea to zero out a buffer in your function -
+  - Use MdePkg.chm to find `ZeroMem` function
+  - Use `ZeroMem` function on your variable buffer "Key" of type EFI_INPUT_KEY
+- Use Boolean flag "ExitLoop" to exit your loop once the user enters a DOT "." character
+
+---
+## Slide 59 Lab 5: How Process (Hints)
+
+**How to create a Feature Flag PCD**
+
+1. Check the MdeModulePkg/MdeModulePkg.dec and search for the "PcdHelloWorldPrintEnable" PCD.
+  - Notice that it is in the \[PcdsFeatureFlag\] section
+2. Add a similar section in the Platform: SimicsOpenBoardPkg/OpenBoard.dec file and add a flag "PcdTypeWriterFeatureEnable" deafult, TRUE. (it can be added to the end of the file)
```

Comment:
2.  Add a similar section in MyPkg,  `edk2/MyPkg/MyPkg.dec` file and add a flag “`PcdTypeWriterFeatureEnable`” default, TRUE. (it can be added to the end of the file)